docs: add data-deal specs for protocol devs (#585)
medibloc/panacea-core: .gitbook/3-protocol-devs/1-dep-specs/2-data-deal.md

# Data Deal

- Status: Draft
- Crated: 2023-01-03
- Modified: 2023-01-03
- Authors
  - Youngjoon Lee <yjlee@medibloc.org>
  - Gyuguen Jang <gyuguen.jang@medibloc.org>
  - Hansol Lee <hansol@medibloc.org>
  - Myongsik Gong <myongsik_gong@medibloc.org>
  - Inchul Song <icsong@medibloc.org>

## Synopsis

This document defines a data deal which is a contract for data collecting and pay for provision in [DEP](../../1-users/3-data-exchange/0-about-dep.md).

> Data consumers create data deal by specifying what type of data and how much they want, and how much they are willing to pay for it.

Reviewer: let's go with these expressions.
https://github.com/medibloc/panacea-core/pull/567#discussion_r1058690047

> Data consumers create data deal by specifying what type of data and how much they want, and how much they are willing to pay for it.

Reviewer: ```suggestion
Data consumers create data deals by specifying what type of data and how much they want, and how much they are willing to pay for it.
```

It's better to use the plural form when you say about a general concept of something.

Data providers can provide their data to the deals that match the type of data they have.

> When a data provider provides data to the deal, the payout is distributed to the provider and oracle that validated the data.

Reviewer: ```suggestion
When a data provider provides data to the deal, the payout is distributed to the provider and an oracle that validated the data.
```

Author: I add `the` instead of `an` because it is being specified which oracle it is (which validated the data).

Also, data consumers can deactivate their data deal whenever they want and the remained budget would be refunded to the consumer's account.

### Motivation

Data consumers want different types of data, and even for the same type of data, they all differ in how much they want and how much they are willing to pay.
Thus, the data deal was devised so that data consumers can determine the type of data they want, as much as they want, at a cost they want.

### Definitions

`Data Provider`, `Data Consumer` and `Oracle` are defined in [User Flow](./1-user-flow.md)

> ## Technical Specification
> 
> ### Data Structure of Deal

Reviewer: ```suggestion
## Technical Specification

Data consumers should be able to post the information described below publicly, so that any data provider can see it. Also, data providers should be assured that a particular data consumer really posted the information.

### Data Structure of Deal
```

The structure of data deal.

```proto
message Deal {
  uint64 id = 1;
  string address = 2;
  repeated string data_schema = 3;
  cosmos.base.v1beta1.Coin budget = 4;
  uint64 max_num_data = 5;
  uint64 cur_num_data = 6;
  string consumer_address = 7;
  DealStatus status = 8;
}
```

- `id`: auto increment id
- `address`: an address of deal generated when deal is created
- `data_schema`: a list of urls of desired type of data schema
- `budget`: budget for data provision

> - `max_num_data`: maximum number of data the consumer want
> - `cur_num_data`: current number of data of the deal

Reviewer: ```suggestion
- `max_num_data`: the maximum number of data the consumer want
- `cur_num_data`: the current number of data of the deal
```

> - `consumer_address`: consumer's account address (panacea1...)

Reviewer: ```suggestion
- `consumer_address`: consumer's account address of the form `panacea1...`
```

> - `address`: an address of deal generated when deal is created
> - `data_schema`: a list of urls of desired type of data schema
> - `budget`: budget for data provision
> - `max_num_data`: maximum number of data the consumer want
> - `cur_num_data`: current number of data of the deal
> - `consumer_address`: consumer's account address (panacea1...)

Reviewer: ```suggestion
- `address`: An address of deal generated when deal is created
- `data_schema`: A list of URLs of desired data schema
- `budget`: A budget for consuming data
- `max_num_data`: The maximum number of data the consumer want
- `cur_num_data`: The current number of data provided
- `consumer_address`: A consumer's account address of the form `panacea1...`
```

- `status`: the status of deal. 3 statuses can be possible
  - `DEAL_STATUS_ACTIVE`: the status when deal is active (`cur_num_data` < `max_num_data`).  
  - `DEAL_STATUS_INACTIVE`: the status when deal is deactivated (when consumer deactivated the deal)
  - `DEAL_STATUS_COMPLETED`: the status when deal is completed (`max_num_data` of data is provided)

### Create Data Deal

Data consumers can create their deal with the followings:

```proto
message MsgCreateDeal {

>   repeated string data_schema = 1;

Reviewer 2: In near future, we can add an example of `data_schema`, too!

Author: Right. But I think it's out of scope of this doc. So It'd be great to add link for more information about json schema.
Do I understand this comment right?

Reviewer 2: Yes, it can be explained details in other docs.

  cosmos.base.v1beta1.Coin budget = 2;
  uint64 max_num_data = 3;
  string consumer_address = 4;
}
```

When deal is created, the amount of budget is sent from consumer's account to deal's account.
In other words, the balance of consumer's account should be greater or equal than the budget.

### Deactivate Data Deal

The consumer who created the deal can deactivate the deal at any time as long as `max_num_data` of data is not provided.

> To deactivate deal, the id of deal should be specified.

Reviewer 2: We can notice that `requester_address` and `consumer_address` must be same.

Author: Good. I'll notice that.

Author: Ah, I wrote `The consumer who created the deal can deactivate ~` in that sense.
Isn't it enough?

Reviewer 2: I think that it's enough

# Data Deal

- Status: Draft

> - Crated: 2023-01-03

Reviewer 2: Typo. Created

Author: thanks
